enhance `CMakeMake` easyblock to check whether correct `Python` installation was picked up by `CMake`

Flamefire opened the pull request:
This is #3233 squashed and rebased to 5.0x

There was 1 commit by Danilo Gonzalez which mostly moved and removed code introduced earlier by  @MartinsNadia which then was removed/changed again in the "revised version".

To make the rebase possible I squashed that together with all commits from @MartinsNadia into a single one which should be fine as the "revised version" was almost a complete rewrite.

Micket reviewed (approved):
lgtm

Micket commented:
One check is just stuck and I see no other way to restart it.

Micket commented:
Test report by @Micket
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _Doxygen-1.11.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _ParaView-5.12.0-foss-2023b.eb_ 
 * **SUCCESS** _igraph-0.10.12-foss-2023b.eb_ 
 * **SUCCESS** _casacore-3.5.0-foss-2023b.eb_ 
 * **SUCCESS** _GDAL-3.9.0-foss-2023b.eb_ 
 * **FAIL (build issue)** _ESPResSo-4.2.2-foss-2023b.eb_ (partial log available at https://gist.github.com/Micket/367fcc43d0801f71a40f13a543aa7993)
 * **FAIL (build issue)** _EveryBeam-0.5.2-foss-2023b.eb_ (partial log available at https://gist.github.com/Micket/2a29246530abd58a00e60b93c260d9f5)
 * **FAIL (build issue)** _WSClean-3.4-foss-2023b.eb_ (partial log available at https://gist.github.com/Micket/46acfa424fe32978f176fda20027e2e3)
 * **FAIL (build issue)** _DP3-6.0-foss-2023b.eb_ (partial log available at https://gist.github.com/Micket/431d2ae59da598d7f8f7060b0e385811)
 * **SUCCESS** _AOFlagger-3.4.0-foss-2023b.eb_ 
 * **SUCCESS** _IDG-1.2.0-foss-2023b.eb_ 

Build succeeded for 7 out of 11 (11 easyconfigs in total)
vera-skylake-build - Linux Rocky Linux 8.9, x86_64, Intel Xeon Processor (Skylake, IBRS, no TSX), Python 3.6.8
See https://gist.github.com/Micket/eb25fe4d270ea3181261c70ed5cd1e69 for a full test report.

boegel commented:
for `EveryBeam-0.5.2-foss-2023b.eb` (required by `WSClean` and `DP3`):
```
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
```

for `ESPResSo-4.2.2-foss-2023b.eb`:
```
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
```

@Micket So in both cases we're just missing `Doxygen` as (build) dependency?

Flamefire commented:
> @Micket So in both cases we're just missing `Doxygen` as (build) dependency?

Yes, although: "The following OPTIONAL packages have not been found: Doxygen", might still be better to have it?

Actual error is related to this PR though:

> Python path errors:
> Python executable does not exist: 1 (at easybuild/easyblocks/generic/cmakemake.py:392 in check_python_paths)

Micket commented on easybuild/easyblocks/generic/cmakemake.py:
```diff
@@ -331,9 +332,61 @@ def configure_step(self, srcdir=None, builddir=None):
+            "include_dir": [],
+            "library": [],
+        }
+        python_regex = re.compile(r"_?(?:Python|PYTHON)\d?_(EXECUTABLE|INCLUDE_DIR|LIBRARY)[^=]*=(.*)")
```
The [^=]* part matches to much. From my test builds, I have things like
```
PYTHON_EXECUTABLE-ADVANCED:INTERNAL=1
```

Micket replied:
maybe sufficient to just add a : ?

```suggetion
        python_regex = re.compile(r"_?(?:Python|PYTHON)\d?_(EXECUTABLE|INCLUDE_DIR|LIBRARY):[^=]*=(.*)")
```

Flamefire replied:
That is not enough to match variables with a potential suffixes like "_DIR" or "_DEBUG", like `PYTHON_LIBRARY_DEBUG`. However limiting to word characters should work. See [the](https://github.com/easybuilders/easybuild-easyblocks/pull/3399/commits/23451116f6395cc34268a594f4ab56ad69cd001a) where I also added some examples in the commit message

boegel commented:
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **FAIL (build issue)** _Arrow-16.1.0-gfbf-2023b.eb_ (partial log available at https://gist.github.com/boegel/5eacad228d09f0cc1f98a5b38221487c)
 * **FAIL (build issue)** _ANTs-2.5.0-foss-2022b.eb_ (partial log available at https://gist.github.com/boegel/332d398796977be08f336353940b7732)

Build succeeded for 0 out of 2 (2 easyconfigs in total)
node3580.doduo.os - Linux RHEL 8.8, x86_64, AMD EPYC 7552 48-Core Processor, Python 3.6.8
See https://gist.github.com/boegel/0fde9883e29c5d8555392095abaf8840 for a full test report.

boegel commented:
@Flamefire Still not there yet, the two easyconfigs in the test report fail with errors like:
```
== FAILED: Installation ended unsuccessfully: Python path errors:
Python executable path does not exist: :INTERNAL
Python include_dir path does not exist: :INTERNAL
Python library path does not exist: :INTERNAL
```

Flamefire commented:
> @Flamefire Still not there yet, the two easyconfigs in the test report fail with errors like:

I added an extra capture group with the last change, so it picked up the type instead of the value...
I removed the non-capturing group which IMO makes the regex easier to read (avoids having `(?::\w+)`) and fixed the indices.

boegel commented:
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ALAMODE-1.4.2-foss-2022b.eb_ 
 * **SUCCESS** _ANTs-2.5.0-foss-2022b.eb_ 
 * **SUCCESS** _AOMP-13.0-2-GCCcore-10.2.0.eb_ 
 * **SUCCESS** _Arcade-Learning-Environment-0.8.1-foss-2023a.eb_ 
 * **SUCCESS** _Arrow-16.1.0-gfbf-2023b.eb_ 
 * **SUCCESS** _awscli-2.11.21-GCCcore-11.3.0.eb_ 
 * **SUCCESS** _btllib-1.7.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _CVXPY-1.4.2-foss-2023a.eb_ 
 * **SUCCESS** _CheckM2-1.0.2-foss-2022b.eb_ 
 * **FAIL (build issue)** _cppyy-3.1.2-GCCcore-13.2.0.eb_ (partial log available at https://gist.github.com/boegel/3f48c98416f098b27fd9951040e7b24e)
 * **SUCCESS** _DIRAC-23.0-intel-2023a.eb_ 
 * **SUCCESS** _Doxygen-1.9.1-GCCcore-10.3.0.eb_ 
 * **SUCCESS** _EGTtools-0.1.11-foss-2022a.eb_ 
 * **SUCCESS** _FLINT-3.1.1-gfbf-2023b.eb_ 
 * **SUCCESS** _FlexiBLAS-3.3.1-GCC-13.2.0.eb_ 
 * **SUCCESS** _flatbuffers-23.1.4-GCCcore-12.2.0.eb_ 
 * **SUCCESS** _fumi_tools-0.18.2-GCC-11.2.0.eb_ 
 * **SUCCESS** _Gymnasium-0.29.1-foss-2023a.eb_ 
 * **SUCCESS** _HepMC3-3.2.6-GCC-12.3.0.eb_ 
 * **SUCCESS** _HiGHS-1.7.0-gfbf-2023b.eb_ 
 * **SUCCESS** _kineto-0.4.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _LLVM-18.1.8-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _MDI-1.4.29-gompi-2023b.eb_ 
 * **SUCCESS** _MEGAHIT-1.2.9-GCCcore-9.3.0.eb_ 
 * **SUCCESS** _MetaBAT-2.15-gompi-2021b.eb_ 
 * **SUCCESS** _ncbi-vdb-3.1.1-gompi-2023b.eb_ 
 * **SUCCESS** _ONNX-1.15.0-gfbf-2023a.eb_ 
 * **SUCCESS** _OpenAI-Gym-0.26.2-foss-2022a.eb_ 
 * **FAIL (build issue)** _OpenCV-4.8.1-foss-2023a-contrib.eb_ (partial log available at https://gist.github.com/boegel/1c9736e79946cb7b75976f39150b0cdc)
 * **SUCCESS** _OpenMEEG-2.5.7-foss-2023a.eb_ 
 * **SUCCESS** _OpenMM-8.0.0-foss-2022a.eb_ 
 * **SUCCESS** _PICI-LIGGGHTS-3.8.1-foss-2022a.eb_ 
 * **SUCCESS** _PyTorch-bundle-2.1.2-foss-2023a.eb_ 
 * **SUCCESS** _pFUnit-4.7.3-gompi-2022a.eb_ 
 * **SUCCESS** _poppler-24.04.0-GCC-13.2.0.eb_ 
 * **SUCCESS** _pybind11-2.9.2-GCCcore-11.3.0.eb_ 
 * **SUCCESS** _pypmt-1.2.0-gfbf-2023a.eb_ 
 * **SUCCESS** _python-mujoco-3.1.4-foss-2023a.eb_ 
 * **SUCCESS** _Rust-1.70.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _SPAdes-3.15.2-GCC-10.2.0-Python-2.7.18.eb_ 
 * **SUCCESS** _SRA-Toolkit-3.0.5-gompi-2022b.eb_ 
 * **SUCCESS** _SentencePiece-0.2.0-GCC-13.2.0.eb_ 
 * **SUCCESS** _SeqAn-2.4.0-GCCcore-9.3.0.eb_ 
 * **SUCCESS** _scikit-build-core-0.5.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _tmap-20220502-GCC-11.2.0.eb_ 
 * **SUCCESS** _UShER-0.5.0-gompi-2021b.eb_ 
 * **SUCCESS** _vcflib-1.0.9-gfbf-2023a-R-4.3.2.eb_ 
 * **SUCCESS** _xtensor-0.24.7-foss-2023a.eb_ 
 * **SUCCESS** _Z3-4.13.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _zlib-ng-2.2.1-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _manta-1.6.0-GCC-10.2.0-Python-2.7.18.eb_ 
 * **SUCCESS** _pybind11-2.7.1-GCCcore-11.2.0-Python-2.7.18.eb_ 

Build succeeded for 50 out of 52 (52 easyconfigs in total)
node3580.doduo.os - Linux RHEL 8.8, x86_64, AMD EPYC 7552 48-Core Processor, Python 3.6.8
See https://gist.github.com/boegel/78d5a4c84c02603bbb22184a55e13e92 for a full test report.

boegel commented:
`cppyy-3.1.2-GCCcore-13.2.0.eb` failed due to an existing lock, I'll upload a new test report for that.

For `OpenCV-4.8.1-foss-2023a-contrib.eb` it seems like we have a regression introduced by the changes here though:
```
== 2024-10-12 22:42:02,342 cmakemake.py:354 INFO Checking Python paths
== 2024-10-12 22:42:02,343 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON2_EXECUTABLE:FILEPATH=
== 2024-10-12 22:42:02,343 cmakemake.py:368 INFO Python executable path: 
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON2_INCLUDE_DIR:PATH=
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python include_dir path: 
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON2_INCLUDE_DIR2:PATH=
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python include_dir path: 
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON2_LIBRARY:FILEPATH=
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python library path: 
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON2_LIBRARY_DEBUG:FILEPATH=
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python library path: 
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON3_EXECUTABLE:FILEPATH=/user/gent/400/vsc40023/eb_arcaninescratch/RHEL8-eb5dev-rpath/zen2-ib/software/Python/3.11.3-GCCcore-12.3.0/bin/python3
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python executable path: /user/gent/400/vsc40023/eb_arcaninescratch/RHEL8-eb5dev-rpath/zen2-ib/software/Python/3.11.3-GCCcore-12.3.0/bin/python3
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON3_INCLUDE_DIR:PATH=/user/gent/400/vsc40023/eb_arcaninescratch/RHEL8-eb5dev-rpath/zen2-ib/software/Python/3.11.3-GCCcore-12.3.0/include/python3.11
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python include_dir path: /user/gent/400/vsc40023/eb_arcaninescratch/RHEL8-eb5dev-rpath/zen2-ib/software/Python/3.11.3-GCCcore-12.3.0/include/python3.11
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON3_INCLUDE_DIR2:PATH=
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python include_dir path: 
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON3_LIBRARY:FILEPATH=/user/gent/400/vsc40023/eb_arcaninescratch/RHEL8-eb5dev-rpath/zen2-ib/software/Python/3.11.3-GCCcore-12.3.0/lib/libpython3.11.so
== 2024-10-12 22:42:02,344 cmakemake.py:368 INFO Python library path: /user/gent/400/vsc40023/eb_arcaninescratch/RHEL8-eb5dev-rpath/zen2-ib/software/Python/3.11.3-GCCcore-12.3.0/lib/libpython3.11.so
== 2024-10-12 22:42:02,344 cmakemake.py:365 DEBUG Python related CMake cache line found: PYTHON3_LIBRARY_DEBUG:FILEPATH=
== 2024-10-12 22:42:02,345 cmakemake.py:368 INFO Python library path: 
== 2024-10-12 22:42:02,346 cmakemake.py:365 DEBUG Python related CMake cache line found: _Python2_EXECUTABLE:INTERNAL=_Python2_EXECUTABLE-NOTFOUND
== 2024-10-12 22:42:02,346 cmakemake.py:368 INFO Python executable path: _Python2_EXECUTABLE-NOTFOUND
== 2024-10-12 22:42:02,353 build_log.py:226 ERROR EasyBuild encountered an error (at easybuild/eb5/easybuild-framework/easybuild/base/exceptions.py:126 in __init__): Python path errors:
Python executable path does not exist: 
Python include_dir path does not exist: 
Python include_dir path does not exist: 
Python include_dir path does not exist: 
Python library path does not exist: 
Python library path does not exist: 
Python library path does not exist:  (at easybuild/easyblocks/generic/cmakemake.py:393 in check_python_paths)

boegel commented:
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _cppyy-3.1.2-GCCcore-13.2.0.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
node3580.doduo.os - Linux RHEL 8.8, x86_64, AMD EPYC 7552 48-Core Processor, Python 3.6.8
See https://gist.github.com/boegel/cbae8ec769b48e72e80aa0e3b0344587 for a full test report.

Flamefire commented:
I didn't account for empty cache entries as they usually won't appear. The find-modules would set them to `*-NOT-FOUND`, not to empty.

It seems that OpenCV [sets those cache entries](https://github.com/opencv/opencv/blob/08f7f13dfab327967f61a6a4242ed233dfeaad32/cmake/OpenCVDetectPython.cmake#L276) even if they are empty.

I went full way and excluded any value considered "false" by CMake which should make it reasonably robust.

We could be more restrictive and use `(Python\d?|PYTHON)` to only include prefixes from known CMake find-modules (`FindPythonInterp` only uses `PYTHON_EXECUTABLE` without any version, the `Python2/3` are the newer `FindPython2/3`) That would exclude the `PYTHON2_EXECUTABLE` set by OpenCV but I guess we do want to check those too (when non-False).

boegel commented:
Test report by @boegel
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _ALAMODE-1.4.2-foss-2022b.eb_ 
 * **SUCCESS** _ANTs-2.5.0-foss-2022b.eb_ 
 * **SUCCESS** _AOMP-13.0-2-GCCcore-10.2.0.eb_ 
 * **SUCCESS** _Arcade-Learning-Environment-0.8.1-foss-2023a.eb_ 
 * **SUCCESS** _Arrow-16.1.0-gfbf-2023b.eb_ 
 * **SUCCESS** _awscli-2.11.21-GCCcore-11.3.0.eb_ 
 * **SUCCESS** _btllib-1.7.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _CVXPY-1.4.2-foss-2023a.eb_ 
 * **SUCCESS** _CheckM2-1.0.2-foss-2022b.eb_ 
 * **SUCCESS** _cppyy-3.1.2-GCCcore-13.2.0.eb_ 
 * **SUCCESS** _DIRAC-23.0-intel-2023a.eb_ 
 * **SUCCESS** _Doxygen-1.9.1-GCCcore-10.3.0.eb_ 
 * **SUCCESS** _EGTtools-0.1.11-foss-2022a.eb_ 
 * **SUCCESS** _FLINT-3.1.1-gfbf-2023b.eb_ 
 * **SUCCESS** _FlexiBLAS-3.3.1-GCC-13.2.0.eb_ 
 * **SUCCESS** _flatbuffers-23.1.4-GCCcore-12.2.0.eb_ 
 * **SUCCESS** _fumi_tools-0.18.2-GCC-11.2.0.eb_ 
 * **SUCCESS** _Gymnasium-0.29.1-foss-2023a.eb_ 
 * **SUCCESS** _HepMC3-3.2.6-GCC-12.3.0.eb_ 
 * **SUCCESS** _HiGHS-1.7.0-gfbf-2023b.eb_ 
 * **SUCCESS** _kineto-0.4.0-GCC-12.3.0.eb_ 
 * **SUCCESS** _LLVM-18.1.8-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _MDI-1.4.29-gompi-2023b.eb_ 
 * **SUCCESS** _MEGAHIT-1.2.9-GCCcore-9.3.0.eb_ 
 * **SUCCESS** _MetaBAT-2.15-gompi-2021b.eb_ 
 * **SUCCESS** _ncbi-vdb-3.1.1-gompi-2023b.eb_ 
 * **SUCCESS** _ONNX-1.15.0-gfbf-2023a.eb_ 
 * **SUCCESS** _OpenAI-Gym-0.26.2-foss-2022a.eb_ 
 * **SUCCESS** _OpenCV-4.8.1-foss-2023a-contrib.eb_ 
 * **SUCCESS** _OpenMEEG-2.5.7-foss-2023a.eb_ 
 * **SUCCESS** _OpenMM-8.0.0-foss-2022a.eb_ 
 * **SUCCESS** _PICI-LIGGGHTS-3.8.1-foss-2022a.eb_ 
 * **SUCCESS** _PyTorch-bundle-2.1.2-foss-2023a.eb_ 
 * **SUCCESS** _pFUnit-4.7.3-gompi-2022a.eb_ 
 * **SUCCESS** _poppler-24.04.0-GCC-13.2.0.eb_ 
 * **SUCCESS** _pybind11-2.9.2-GCCcore-11.3.0.eb_ 
 * **SUCCESS** _pypmt-1.2.0-gfbf-2023a.eb_ 
 * **SUCCESS** _python-mujoco-3.1.4-foss-2023a.eb_ 
 * **SUCCESS** _Rust-1.70.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _SPAdes-3.15.2-GCC-10.2.0-Python-2.7.18.eb_ 
 * **SUCCESS** _SRA-Toolkit-3.0.5-gompi-2022b.eb_ 
 * **SUCCESS** _SentencePiece-0.2.0-GCC-13.2.0.eb_ 
 * **SUCCESS** _SeqAn-2.4.0-GCCcore-9.3.0.eb_ 
 * **SUCCESS** _scikit-build-core-0.5.0-GCCcore-12.3.0.eb_ 
 * **SUCCESS** _tmap-20220502-GCC-11.2.0.eb_ 
 * **SUCCESS** _UShER-0.5.0-gompi-2021b.eb_ 
 * **SUCCESS** _vcflib-1.0.9-gfbf-2023a-R-4.3.2.eb_ 
 * **SUCCESS** _xtensor-0.24.7-foss-2023a.eb_ 
 * **SUCCESS** _Z3-4.13.0-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _zlib-ng-2.2.1-GCCcore-13.3.0.eb_ 
 * **SUCCESS** _manta-1.6.0-GCC-10.2.0-Python-2.7.18.eb_ 
 * **SUCCESS** _pybind11-2.7.1-GCCcore-11.2.0-Python-2.7.18.eb_ 

Build succeeded for 52 out of 52 (52 easyconfigs in total)
node3508.doduo.os - Linux RHEL 8.8, x86_64, AMD EPYC 7552 48-Core Processor, Python 3.6.8
See https://gist.github.com/boegel/d9cf61969f78c4c90ae98271f3c80b2e for a full test report.

boegel reviewed (approved):
lgtm
